Mathlib/CategoryTheory/Sites/Coherent.lean

```diff
@@ -0,0 +1,81 @@
+/-
+Copyright (c) 2023 Adam Topaz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adam Topaz
+-/
+
+import Mathlib.CategoryTheory.Sites.Coverage
+import Mathlib.CategoryTheory.Sites.EffectiveEpimorphic
+
+/-!
+
+# The Coherent Grothendieck Topology
+
+This file defines the coherent Grothendieck topology (and coverage) on a category `C`.
+The category `C` must satisfy a `Precoherent C` condition, which is essentially the minimal
+requirement for the coherent coverage to exist.
+
+In `isSheaf_coherent`, we characterize the sheaf condition for presheaves of types for the
+coherent Grothendieck topology in terms of finite effective epimorphic families.
+
```

Comment:
Perhaps you can add some hint to the literature at where the notion of "effective" is provided, explaining why you restrict to `Fintype α`?

Reply:
As far as I know, we don't have a definition of `subcanonical` yet.

It's a bit difficult to add references, because they all assume coherentness in order to define the coherent topology. The `Precoherent` condition is something I came up with just to be able to define the coherent topology. Under other assumptions (like regularity) the notion of effective epimorphic family can be simplified significantly, but it seems that the right general notion is what I define in the other file in this PR, as that agrees with the sieve-theoretic definition (as proven in that file).

I guess it would be worthwhile to add a reference to the elephant in the effective epi file, as there the notion of effective epimorphic sieves are mentioned.

Reply:
As far as I know, we don't have a definition of `subcanonical` yet.

It's a bit difficult to add references, because they all assume coherentness in order to define the coherent topology. The `Precoherent` condition is something I came up with just to be able to define the coherent topology. Under other assumptions (like regularity) the notion of effective epimorphic family can be simplified significantly, but it seems that the right general notion is what I define in the other file in this PR, as that agrees with the sieve-theoretic definition (as proven in that file).

I guess it would be worthwhile to add a reference to the elephant in the effective epi file, as there the notion of effective epimorphic sieves are mentioned. I will do that soon.

Reply:
Done!